torch/csrc/cuda/nccl.cpp

```diff
@@ -0,0 +1,175 @@
+    if (input.get_device() != output.get_device()) {
+      throw std::runtime_error("input and output must be on the same device");
+    }
+
+    // all inputs must be same size
+    if (input.numel() != numel) {
+      throw std::runtime_error("all inputs must have the same number of elements");
+    }
+  
+    // outputs have to be of size * size_multiplier
+    if (output.numel() != (numel * size_multiplier)) {
+      throw std::runtime_error("output must be of size input_size * size_multiplier");
+    }
+  }
+}
+
+static ncclDataType_t _get_data_type(at::TypeID type) {
+  switch(type) {
+  case at::TypeID::CUDAFloat   : return ncclFloat;
+  case at::TypeID::CUDAHalf    : return ncclHalf;
+  case at::TypeID::CUDADouble  : return ncclDouble;
+  case at::TypeID::CUDALong    : return ncclInt64;
+  case at::TypeID::CUDAInt     : return ncclInt;
+  case at::TypeID::CUDAChar    : return ncclChar;
+  case at::TypeID::CUDAByte    : return ncclChar;
+  default: throw std::runtime_error("Unconvertible NCCL type");
+  }
+}
+
+PyObject * THCPModule_nccl_reduce(PyObject *self, PyObject *args) {
+  HANDLE_TH_ERRORS
+  PyObject *_inputs, *_outputs, *_streams;
+  int root, op, original_device;
+  int64_t count;
+  ncclDataType_t data_type;
+  ncclComm_t* comm;
+  std::mutex* mutex;
+
+  if (!PyArg_ParseTuple(args, "OOOii", &_inputs, &_outputs, &_streams, &root, &op)) {
+    THPUtils_invalidArguments(args, NULL, "nccl_reduce", 1,
+			      "(sequence[Tensor] inputs, sequence[Tensor]"
+			      " outputs, sequence[torch.cuda.Stream or None], int root, int op");
+    return NULL;
+  }
+
+  THPUtils_assert(PySequence_Check(_inputs), "inputs has to be a PySequence");
+  THPUtils_assert(PySequence_Check(_outputs), "outputs has to be a PySequence");
+  THPUtils_assert(PySequence_Check(_streams), "streams has to be a PySequence");
+
+  std::vector<at::Tensor> inputs = THPUtils_PySequence_to_TensorList(THPObjectPtr(PySequence_Fast(_inputs, NULL)));
+  std::vector<at::Tensor> outputs = THPUtils_PySequence_to_TensorList(THPObjectPtr(PySequence_Fast(_outputs, NULL)));
+  std::vector<THCStream*> streams = THPUtils_PySequence_to_THCStreamList(THPObjectPtr(PySequence_Fast(_streams, NULL)));
+
+  THPUtils_assert(inputs.size() == streams.size(), "number of streams is not equal to number of inputs");
+  
+  // we can safely release GIL after this line, no python API used
+  AutoNoGIL no_gil;
+  _check_inputs(inputs, outputs, 1);
+  size_t len = inputs.size();
+
+  data_type = _get_data_type(inputs[0].type().ID());
+  comm = _get_communicator(inputs, len);
+  THCudaCheck(cudaGetDevice(&original_device));
+  count = inputs[0].numel();
+  mutex = THCCachingAllocator_getCudaFreeMutex();
+  mutex->lock();
+#if defined(NCCL_MAJOR) && (NCCL_MAJOR >= 2)
+  ncclGroupStart();
+#endif
+  for (size_t i = 0; i < len; i++) {
+    int device = inputs[i].get_device();
+    if (device != original_device) {
+      THCudaCheck(cudaSetDevice(device));
+    }
```

Comment:
This checks seems incorrect. Consider a case where original device is 2, and you get tensors on devices [0, 2, 4]. You will set the device to 0, because it's different, then you'll move on to tensor on device 2, and it will be equal to original device, so you won't do anything, even though you're on an incorrect one!

Reply:
good catch, will fix.